src/main/scala/SerialAdapter.scala

```diff
@@ -208,38 +226,117 @@ class SimSerial(w: Int) extends BlackBox with HasBlackBoxResource {
   addResource("/testchipip/csrc/SimSerial.cc")
 }
 
-
-case object SerialKey extends Field[Boolean](false)
-case class SerialAdapterParams(masterWhere: TLBusWrapperLocation = FBUS)
-case object SerialAttachKey extends Field[SerialAdapterParams](SerialAdapterParams())
-
-trait CanHavePeripherySerial extends HasPeripheryDebug { this: BaseSubsystem =>
-  private val portName = "serial-adapter"
-
-  val serial = if (p(SerialKey)) {
-    val tlbus = locateTLBusWrapper(p(SerialAttachKey).masterWhere)
+case class SerialTLParams(memParams: MasterPortParams, isMemoryDevice: Boolean = false, width: Int = 4)
+case object SerialTLKey extends Field[Option[SerialTLParams]](None)
+
+case class SerialTLAttachParams(
+  masterWhere: TLBusWrapperLocation = FBUS,
+  slaveWhere: TLBusWrapperLocation = MBUS
+)
+case object SerialTLAttachKey extends Field[SerialTLAttachParams](SerialTLAttachParams())
+
+
+trait CanHavePeripheryTLSerial { this: BaseSubsystem =>
+  private val portName = "serial-tl"
+  val (serdesser, serial_tl) = p(SerialTLKey).map { params =>
+    val memParams = params.memParams
+    val manager = locateTLBusWrapper(p(SerialTLAttachKey).slaveWhere) // The bus for which this acts as a manager
+    val client = locateTLBusWrapper(p(SerialTLAttachKey).masterWhere) // The bus for which this acts as a client
+    val device = if (params.isMemoryDevice) new MemoryDevice else new SimpleDevice("lbwif-ram", Nil)
+    val clientPortParams = TLMasterPortParameters.v1(
+      clients = Seq(TLMasterParameters.v1(
+        name = "serial-tl",
+        sourceId = IdRange(0, 1)
+      ))
+    )
+    val managerPortParams = TLSlavePortParameters.v1(
+      managers = Seq(TLSlaveParameters.v1(
+        address            = AddressSet.misaligned(memParams.base, memParams.size),
+        resources          = device.reg,
+        regionType         = RegionType.UNCACHED, // cacheable
+        executable         = true,
+        supportsGet        = TransferSizes(1, manager.blockBytes),
+        supportsPutFull    = TransferSizes(1, manager.blockBytes),
+        supportsPutPartial = TransferSizes(1, manager.blockBytes))),
+      beatBytes = memParams.beatBytes
+    )
 
     val domain = LazyModule(new ClockSinkDomain(name=Some(portName)))
-    domain.clockNode := tlbus.fixedClockNode
-
-    val inner_io = domain {
-      val adapter = LazyModule(new SerialAdapter)
-      tlbus.fromPort(Some(portName))() := adapter.node
-      InModuleBody {
-        val inner_io = IO(new SerialIO(SERIAL_IF_WIDTH)).suggestName("serial")
-        inner_io.out <> Queue(adapter.module.io.serial.out)
-        adapter.module.io.serial.in <> Queue(inner_io.in)
-        inner_io
-      }
+
+    // TODO: currently assume manager and client buses have same clock
```

Comment:
We should also file an issue to fix this and add a programmable divider on this path. The on-chip busses can usually run much faster than this simple off-chip interface.

Reply:
This seems like a good use case to test https://github.com/ucb-bar/chipyard/pull/676

Reply:
I think that could be an option but I think we also want a purely synthesizable programmable divider like the one found in testchipip currently.